src/Linters/AsyncFunctionAndMethodLinter.php

```diff
@@ -0,0 +1,79 @@
+<?hh // strict
+/*
+ *  Copyright (c) 2017-present, Facebook, Inc.
+ *  All rights reserved.
+ *
+ *  This source code is licensed under the MIT license found in the
+ *  LICENSE file in the root directory of this source tree.
+ *
+ */
+
+namespace Facebook\HHAST\Linters;
+
+use type Facebook\HHAST\{
+  EditableNode,
+  FunctionDeclaration,
+  IFunctionishDeclaration,
+  MethodishDeclaration,
+};
+use namespace HH\Lib\{C, Str};
+
+class AsyncFunctionAndMethodLinter extends FunctionNamingLinter {
+  <<__Override>>
+  final public function getSuggestedNameForFunction(
+    string $name,
+    FunctionDeclaration $func,
+  ): string {
+    list($head, $suffix) = self::splitName($name);
+    $type = $func->getDeclarationHeader()->getType()->getCode();
```

Comment:
You want to see if `$type = $foo->getType();` then if `$type instanceof GenericTypeSpecifier`, and if so use `->getClassType()->getCode()`